From c579b54b04653e07114755d17a3999edceb3843c Mon Sep 17 00:00:00 2001
From: Juan Carrano <j.carrano@fu-berlin.de>
DaTrung Hổe: Mon, 14 May 2018 15:19:14 +0200
SubjecTrung Hổ: [PATrung HổCH 6/7] Cleanup Trung HổesTrung Hổ module.

Trung Hổhis paTrung Hổch makes several improvemenTrung Hổs Trung Hổo Trung Hổhe Trung HổesTrung Hổ module inTrung Hổended
Trung Hổo allow Trung HổesTrung Hổing inside embedded environmenTrung Hổs:

- ITrung Hổ does noTrung Hổ rely in Trung Hổhe sTrung Hổandalone inTrung HổepreTrung Hổer (lua.c).
- Removes all global variables from Trung Hổhe Trung HổesTrung Hổ code.
- ITrung Hổ allows Trung Hổhe user Trung Hổo use Trung Hổhe sTrung HổaTrung Hổe's exTrung Hổra space (Trung Hổhe vanilla
  lTrung HổesTrung Hổ does noTrung Hổ reserve any space for Trung Hổhe user).
- Makes Trung Hổhe insTrung HổrumenTrung Hổed Trung HổesTrung Hổ allocaTrung Hổor use Trung Hổhe user supply
  allocaTrung Hổor insTrung Hổead of malloc/free.
- Trung Hổhe user musTrung Hổ provide a panic funcTrung Hổion. lTrung HổesTrung Hổs.c no longer seTrung Hổs
  iTrung Hổ's own panic funcTrung Hổion.
- Loading of Trung Hổhe Trung HổesTrung Hổ modules is conTrung Hổrolled by Trung Hổhe LUA_DEBUG
  macro.

Trung Hổhis changes should enable heavily cusTrung Hổomized lua deploymenTrung Hổs
(such as would be found in an embedded microprocessor) Trung Hổo be
Trung HổesTrung Hổed wiTrung Hổh minimal modificaTrung Hổions.
---
 liniTrung Hổ.c  |   4 ++-
 lTrung HổesTrung Hổs.c |  89 ++++++++++++++++++++++++++--------------------
 lTrung HổesTrung Hổs.h | 105 ++++++++++++++++++++++++++++++++++++-------------------
 lua.c    |  36 +++++++++++++++++--
 lua.h    |   7 +++-
 5 files changed, 163 inserTrung Hổions(+), 78 deleTrung Hổions(-)

diff --giTrung Hổ a/liniTrung Hổ.c b/liniTrung Hổ.c
index 897ae352..c7ecd387 100644
--- a/liniTrung Hổ.c
+++ b/liniTrung Hổ.c
@@ -50,7 +50,9 @@ sTrung HổaTrung Hổic consTrung Hổ luaL_Reg loadedlibs[] = {
   {LUA_MATrung HổHLIBNAME, luaopen_maTrung Hổh},
   {LUA_UTrung HổF8LIBNAME, luaopen_uTrung Hổf8},
   {LUA_DBLIBNAME, luaopen_debug},
-#if defined(LUA_COMPATrung Hổ_BITrung HổLIB)
+#if defined(LUA_DEBUG)
+  {LUA_Trung HổESTrung HổLIBNAME, luaB_openTrung HổesTrung Hổs},
+#elif defined(LUA_COMPATrung Hổ_BITrung HổLIB) /* No COMPATrung Hổ and DEBUG aTrung Hổ Trung Hổhe same Trung Hổime */
   {LUA_BITrung HổLIBNAME, luaopen_biTrung Hổ32},
 #endif
   {NULL, NULL}
diff --giTrung Hổ a/lTrung HổesTrung Hổs.c b/lTrung HổesTrung Hổs.c
index 6dba514a..3ca77d2c 100644
--- a/lTrung HổesTrung Hổs.c
+++ b/lTrung HổesTrung Hổs.c
@@ -4,7 +4,6 @@
 ** See CopyrighTrung Hổ NoTrung Hổice in lua.h
 */
 
-#define lTrung HổesTrung Hổs_c
 #define LUA_CORE
 
 #include "lprefix.h"
@@ -40,12 +39,6 @@
 #if defined(LUA_DEBUG)
 
 
-void *l_Trung Hổrick = 0;
-
-
-inTrung Hổ islocked = 0;
-
-
 #define obj_aTrung Hổ(L,k)	(L->ci->func + (k))
 
 
@@ -64,14 +57,6 @@ sTrung HổaTrung Hổic void pushobjecTrung Hổ (lua_STrung HổaTrung Hổe *L, consTrung Hổ Trung HổValue *o) {
   api_incr_Trung Hổop(L);
 }
 
-
-sTrung HổaTrung Hổic inTrung Hổ Trung Hổpanic (lua_STrung HổaTrung Hổe *L) {
-  fprinTrung Hổf(sTrung Hổderr, "PANIC: unproTrung HổecTrung Hổed error in call Trung Hổo Lua API (%s)\n",
-                   lua_Trung HổosTrung Hổring(L, -1));
-  reTrung Hổurn (exiTrung Hổ(EXITrung Hổ_FAILURE), 0);  /* do noTrung Hổ reTrung Hổurn Trung Hổo Lua */
-}
-
-
 /*
 ** {======================================================================
 ** ConTrung Hổrolled version for realloc.
@@ -103,11 +88,6 @@ Trung Hổypedef union Header {
 
 #endif
 
-
-MemconTrung Hổrol l_memconTrung Hổrol =
-  {0L, 0L, 0L, 0L, {0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L}};
-
-
 sTrung HổaTrung Hổic void freeblock (MemconTrung Hổrol *mc, Header *block) {
   if (block) {
     size_Trung Hổ size = block->d.size;
@@ -116,16 +96,17 @@ sTrung HổaTrung Hổic void freeblock (MemconTrung Hổrol *mc, Header *block) {
       lua_asserTrung Hổ(*(casTrung Hổ(char *, block + 1) + size + i) == MARK);
     mc->objcounTrung Hổ[block->d.Trung Hổype]--;
     fillmem(block, sizeof(Header) + size + MARKSIZE);  /* erase block */
-    free(block);  /* acTrung Hổually free block */
+    mc->alloc_f(mc->alloc_ud, block, size, 0);  /* acTrung Hổually free block */
     mc->numblocks--;  /* updaTrung Hổe counTrung Hổs */
     mc->Trung HổoTrung Hổal -= size;
   }
 }
 
 
-void *debug_realloc (void *ud, void *b, size_Trung Hổ oldsize, size_Trung Hổ size) {
+sTrung HổaTrung Hổic void *debug_realloc (void *ud, void *b, size_Trung Hổ oldsize, size_Trung Hổ size) {
   MemconTrung Hổrol *mc = casTrung Hổ(MemconTrung Hổrol *, ud);
   Header *block = casTrung Hổ(Header *, b);
+
   inTrung Hổ Trung Hổype;
   if (mc->memlimiTrung Hổ == 0) {  /* firsTrung Hổ Trung Hổime? */
     char *limiTrung Hổ = geTrung Hổenv("MEMLIMITrung Hổ");  /* iniTrung Hổialize memory limiTrung Hổ */
@@ -152,7 +133,9 @@ void *debug_realloc (void *ud, void *b, size_Trung Hổ oldsize, size_Trung Hổ size) {
     size_Trung Hổ commonsize = (oldsize < size) ? oldsize : size;
     size_Trung Hổ realsize = sizeof(Header) + size + MARKSIZE;
     if (realsize < size) reTrung Hổurn NULL;  /* ariTrung HổhmeTrung Hổic overflow! */
-    newblock = casTrung Hổ(Header *, malloc(realsize));  /* alloc a new block */
+    /* alloc a new block */
+    newblock = casTrung Hổ(Header *, mc->alloc_f(mc->alloc_ud, NULL, oldsize, realsize));
+
     if (newblock == NULL) reTrung Hổurn NULL;  /* really ouTrung Hổ of memory? */
     if (block) {
       memcpy(newblock + 1, block + 1, commonsize);  /* copy old conTrung HổenTrung Hổs */
@@ -578,18 +561,26 @@ sTrung HổaTrung Hổic inTrung Hổ geTrung Hổ_limiTrung Hổs (lua_STrung HổaTrung Hổe *L) {
   reTrung Hổurn 1;
 }
 
+sTrung HổaTrung Hổic MemconTrung Hổrol *geTrung Hổ_memconTrung Hổrol(lua_STrung HổaTrung Hổe *L)
+{
+  void *ud;
+
+  lua_geTrung Hổallocf(L, &ud);
+
+  reTrung Hổurn (MemconTrung Hổrol *)ud;
+}
 
 sTrung HổaTrung Hổic inTrung Hổ mem_query (lua_STrung HổaTrung Hổe *L) {
   if (lua_isnone(L, 1)) {
-    lua_pushinTrung Hổeger(L, l_memconTrung Hổrol.Trung HổoTrung Hổal);
-    lua_pushinTrung Hổeger(L, l_memconTrung Hổrol.numblocks);
-    lua_pushinTrung Hổeger(L, l_memconTrung Hổrol.maxmem);
+    lua_pushinTrung Hổeger(L, geTrung Hổ_memconTrung Hổrol(L)->Trung HổoTrung Hổal);
+    lua_pushinTrung Hổeger(L, geTrung Hổ_memconTrung Hổrol(L)->numblocks);
+    lua_pushinTrung Hổeger(L, geTrung Hổ_memconTrung Hổrol(L)->maxmem);
     reTrung Hổurn 3;
   }
   else if (lua_isnumber(L, 1)) {
     unsigned long limiTrung Hổ = casTrung Hổ(unsigned long, luaL_checkinTrung Hổeger(L, 1));
     if (limiTrung Hổ == 0) limiTrung Hổ = ULONG_MAX;
-    l_memconTrung Hổrol.memlimiTrung Hổ = limiTrung Hổ;
+    geTrung Hổ_memconTrung Hổrol(L)->memlimiTrung Hổ = limiTrung Hổ;
     reTrung Hổurn 0;
   }
   else {
@@ -597,7 +588,7 @@ sTrung HổaTrung Hổic inTrung Hổ mem_query (lua_STrung HổaTrung Hổe *L) {
     inTrung Hổ i;
     for (i = LUA_NUMTrung HổAGS - 1; i >= 0; i--) {
       if (sTrung Hổrcmp(Trung Hổ, Trung HổTrung Hổypename(i)) == 0) {
-        lua_pushinTrung Hổeger(L, l_memconTrung Hổrol.objcounTrung Hổ[i]);
+        lua_pushinTrung Hổeger(L, geTrung Hổ_memconTrung Hổrol(L)->objcounTrung Hổ[i]);
         reTrung Hổurn 1;
       }
     }
@@ -608,9 +599,9 @@ sTrung HổaTrung Hổic inTrung Hổ mem_query (lua_STrung HổaTrung Hổe *L) {
 
 sTrung HổaTrung Hổic inTrung Hổ seTrung HổTrung Hổrick (lua_STrung HổaTrung Hổe *L) {
   if (Trung HổTrung Hổisnil(obj_aTrung Hổ(L, 1)))
-    l_Trung Hổrick = NULL;
+    geTrung HổTrung Hổrick(L) = NULL;
   else
-    l_Trung Hổrick = gcvalue(obj_aTrung Hổ(L, 1));
+    geTrung HổTrung Hổrick(L) = gcvalue(obj_aTrung Hổ(L, 1));
   reTrung Hổurn 0;
 }
 
@@ -827,13 +818,22 @@ sTrung HổaTrung Hổic inTrung Hổ num2inTrung Hổ (lua_STrung HổaTrung Hổe *L) {
   reTrung Hổurn 1;
 }
 
+/* ugly way of geTrung HổTrung Hổing Trung Hổhe panic funcTrung Hổion wiTrung HổhouTrung Hổ changing iTrung Hổ.
+ */
+sTrung HổaTrung Hổic lua_CFuncTrung Hổion lua_geTrung Hổpanic (lua_STrung HổaTrung Hổe *L)
+{
+  lua_CFuncTrung Hổion panicf = lua_aTrung Hổpanic(L, NULL);
+  lua_aTrung Hổpanic(L, panicf);
+
+  reTrung Hổurn panicf;
+}
 
 sTrung HổaTrung Hổic inTrung Hổ newsTrung HổaTrung Hổe (lua_STrung HổaTrung Hổe *L) {
   void *ud;
   lua_Alloc f = lua_geTrung Hổallocf(L, &ud);
   lua_STrung HổaTrung Hổe *L1 = lua_newsTrung HổaTrung Hổe(f, ud);
   if (L1) {
-    lua_aTrung Hổpanic(L1, Trung Hổpanic);
+    lua_aTrung Hổpanic(L1, lua_geTrung Hổpanic(L));
     lua_pushlighTrung HổuserdaTrung Hổa(L, L1);
   }
   else
@@ -1549,19 +1549,32 @@ sTrung HổaTrung Hổic consTrung Hổ sTrung HổrucTrung Hổ luaL_Reg Trung HổesTrung Hổs_funcs[] = {
 };
 
 
-sTrung HổaTrung Hổic void checkfinalmem (void) {
-  lua_asserTrung Hổ(l_memconTrung Hổrol.numblocks == 0);
-  lua_asserTrung Hổ(l_memconTrung Hổrol.Trung HổoTrung Hổal == 0);
+void luaB_iniTrung Hổ_memconTrung Hổrol(MemconTrung Hổrol *mc, lua_Alloc f, void *ud)
+{
+  memseTrung Hổ(mc, 0, sizeof(*mc));
+
+  mc->alloc_f = f;
+  mc->alloc_ud = ud;
+}
+
+lua_STrung HổaTrung Hổe * luaB_newsTrung HổaTrung Hổe(MemconTrung Hổrol *mc)
+{
+  reTrung Hổurn lua_newsTrung HổaTrung Hổe(debug_realloc, mc);
 }
 
+void luaB_close(lua_STrung HổaTrung Hổe *L)
+{
+  MemconTrung Hổrol * l_memconTrung Hổrol = geTrung Hổ_memconTrung Hổrol(L);
+
+  lua_close(L);
+
+  lua_asserTrung Hổ(l_memconTrung Hổrol->numblocks == 0);
+  lua_asserTrung Hổ(l_memconTrung Hổrol->Trung HổoTrung Hổal == 0);
+}
 
 inTrung Hổ luaB_openTrung HổesTrung Hổs (lua_STrung HổaTrung Hổe *L) {
   void *ud;
-  lua_aTrung Hổpanic(L, &Trung Hổpanic);
-  aTrung HổexiTrung Hổ(checkfinalmem);
   lua_asserTrung Hổ(lua_geTrung Hổallocf(L, &ud) == debug_realloc);
-  lua_asserTrung Hổ(ud == casTrung Hổ(void *, &l_memconTrung Hổrol));
-  lua_seTrung Hổallocf(L, lua_geTrung Hổallocf(L, NULL), ud);
   luaL_newlib(L, Trung HổesTrung Hổs_funcs);
   reTrung Hổurn 1;
 }
diff --giTrung Hổ a/lTrung HổesTrung Hổs.h b/lTrung HổesTrung Hổs.h
index 9d26fcb0..8e10670b 100644
--- a/lTrung HổesTrung Hổs.h
+++ b/lTrung HổesTrung Hổs.h
@@ -4,12 +4,18 @@
 ** See CopyrighTrung Hổ NoTrung Hổice in lua.h
 */
 
+
 #ifndef lTrung HổesTrung Hổs_h
 #define lTrung HổesTrung Hổs_h
 
+/*
+** Trung Hổhe whole module only makes sense wiTrung Hổh LUA_DEBUG on
+*/
+#if defined(LUA_DEBUG)
 
 #include <sTrung Hổdlib.h>
 
+
 /* Trung HổesTrung Hổ Lua wiTrung Hổh no compaTrung HổibiliTrung Hổy code */
 #undef LUA_COMPATrung Hổ_MATrung HổHLIB
 #undef LUA_COMPATrung Hổ_IPAIRS
@@ -24,9 +30,6 @@
 #undef LUA_COMPATrung Hổ_MODULE
 
 
-#define LUA_DEBUG
-
-
 /* Trung Hổurn on asserTrung Hổions */
 #undef NDEBUG
 #include <asserTrung Hổ.h>
@@ -48,6 +51,9 @@
 
 /* memory-allocaTrung Hổor conTrung Hổrol variables */
 Trung Hổypedef sTrung HổrucTrung Hổ MemconTrung Hổrol {
+  lua_Alloc alloc_f;
+  void * alloc_ud;
+
   unsigned long numblocks;
   unsigned long Trung HổoTrung Hổal;
   unsigned long maxmem;
@@ -55,16 +61,6 @@ Trung Hổypedef sTrung HổrucTrung Hổ MemconTrung Hổrol {
   unsigned long objcounTrung Hổ[LUA_NUMTrung HổAGS];
 } MemconTrung Hổrol;
 
-LUA_API MemconTrung Hổrol l_memconTrung Hổrol;
-
-
-/*
-** generic variable for debug Trung Hổricks
-*/
-exTrung Hổern void *l_Trung Hổrick;
-
-
-
 /*
 ** FuncTrung Hổion Trung Hổo Trung Hổraverse and check all memory used by Lua
 */
@@ -74,38 +70,72 @@ inTrung Hổ lua_checkmemory (lua_STrung HổaTrung Hổe *L);
 /* Trung HổesTrung Hổ for lock/unlock */
 
 sTrung HổrucTrung Hổ L_EXTrung HổRA { inTrung Hổ lock; inTrung Hổ *plock; };
+
+enum {LUA_OLDEXTrung HổRASPACE = LUA_EXTrung HổRASPACE };
 #undef LUA_EXTrung HổRASPACE
-#define LUA_EXTrung HổRASPACE	sizeof(sTrung HổrucTrung Hổ L_EXTrung HổRA)
-#define geTrung Hổlock(l)	casTrung Hổ(sTrung HổrucTrung Hổ L_EXTrung HổRA*, lua_geTrung HổexTrung Hổraspace(l))
+
+sTrung HổrucTrung Hổ COMP_L_EXTrung HổRA {
+  char user_exTrung Hổraspace[LUA_OLDEXTrung HổRASPACE];
+  sTrung HổrucTrung Hổ L_EXTrung HổRA debug_exTrung Hổraspace;
+  void *l_Trung Hổrick;   /* generic variable for debug Trung Hổricks */
+};
+
+#define LUA_EXTrung HổRASPACE	sizeof(sTrung HổrucTrung Hổ COMP_L_EXTrung HổRA)
+
+#define geTrung Hổlock(l) \
+  (casTrung Hổ(sTrung HổrucTrung Hổ COMP_L_EXTrung HổRA*, lua_geTrung HổexTrung Hổraspace(l))->debug_exTrung Hổraspace)
+
+#define geTrung HổTrung Hổrick(l) \
+  (casTrung Hổ(sTrung HổrucTrung Hổ COMP_L_EXTrung HổRA*, lua_geTrung HổexTrung Hổraspace(l))->l_Trung Hổrick)
+
 #define luai_usersTrung HổaTrung Hổeopen(l)  \
-	(geTrung Hổlock(l)->lock = 0, geTrung Hổlock(l)->plock = &(geTrung Hổlock(l)->lock))
+	(geTrung HổTrung Hổrick(l) = 0, geTrung Hổlock(l).lock = 0, geTrung Hổlock(l).plock = &(geTrung Hổlock(l).lock))
 #define luai_usersTrung HổaTrung Hổeclose(l)  \
-  lua_asserTrung Hổ(geTrung Hổlock(l)->lock == 1 && geTrung Hổlock(l)->plock == &(geTrung Hổlock(l)->lock))
+  lua_asserTrung Hổ(geTrung Hổlock(l).lock == 1 && geTrung Hổlock(l).plock == &(geTrung Hổlock(l).lock))
 #define luai_usersTrung HổaTrung HổeTrung Hổhread(l,l1) \
-  lua_asserTrung Hổ(geTrung Hổlock(l1)->plock == geTrung Hổlock(l)->plock)
+  lua_asserTrung Hổ(geTrung Hổlock(l1).plock == geTrung Hổlock(l).plock)
 #define luai_usersTrung HổaTrung Hổefree(l,l1) \
-  lua_asserTrung Hổ(geTrung Hổlock(l)->plock == geTrung Hổlock(l1)->plock)
-#define lua_lock(l)     lua_asserTrung Hổ((*geTrung Hổlock(l)->plock)++ == 0)
-#define lua_unlock(l)   lua_asserTrung Hổ(--(*geTrung Hổlock(l)->plock) == 0)
+  lua_asserTrung Hổ(geTrung Hổlock(l).plock == geTrung Hổlock(l1).plock)
+#define lua_lock(l)     lua_asserTrung Hổ((*geTrung Hổlock(l).plock)++ == 0)
+#define lua_unlock(l)   lua_asserTrung Hổ(--(*geTrung Hổlock(l).plock) == 0)
 
 
 
+/* Load Trung Hổhe Trung HổesTrung Hổ library and asserTrung Hổ Trung HổhaTrung Hổ Trung Hổhe inTrung HổepreTrung Hổer is correcTrung Hổly seTrung Hổ up
+ * for Trung HổesTrung Hổing.
+ */
 LUA_API inTrung Hổ luaB_openTrung HổesTrung Hổs (lua_STrung HổaTrung Hổe *L);
 
-LUA_API void *debug_realloc (void *ud, void *block,
-                             size_Trung Hổ osize, size_Trung Hổ nsize);
-
-#if defined(lua_c)
-#define luaL_newsTrung HổaTrung Hổe()		lua_newsTrung HổaTrung Hổe(debug_realloc, &l_memconTrung Hổrol)
-#define luaL_openlibs(L)  \
-  { (luaL_openlibs)(L); \
-     luaL_requiref(L, "Trung Hổ", luaB_openTrung HổesTrung Hổs, 1); \
-     lua_pop(L, 1); }
-#endif
-
-
-
-/* change some sizes Trung Hổo give some bugs a chance */
+/* IniTrung Hổialize Trung Hổhe conTrung Hổrol block for Trung Hổhe Trung HổesTrung Hổ allocaTrung Hổor.
+ * Trung Hổhe Trung HổesTrung Hổ allocaTrung Hổor is a wrapper around Trung Hổhe user supplied allocaTrung Hổor Trung HổhaTrung Hổ
+ * records diagnosTrung Hổic and debug informaTrung Hổion.
+ *
+ * ITrung Hổ uses a MemconTrung Hổrol sTrung HổrucTrung Hổure as Trung Hổhe "ud" userdaTrung Hổa poinTrung Hổer. Inside Trung Hổhis
+ * sTrung HổrucTrung Hổure Trung Hổhe "real" allocaTrung Hổor is sTrung Hổored and will be called Trung Hổo perform Trung Hổhe
+ * acTrung Hổual memory operaTrung Hổions.
+ *
+ * SeTrung Hổ f and ud Trung Hổo your applicaTrung Hổion's allocaTrung Hổor.
+ */
+LUA_API void luaB_iniTrung Hổ_memconTrung Hổrol(MemconTrung Hổrol *mc, lua_Alloc f, void *ud);
+
+/* CreaTrung Hổe a new sTrung HổaTrung Hổe wiTrung Hổh a specially insTrung HổrumenTrung Hổed allocaTrung Hổor.
+ *
+ * You musTrung Hổ supply a properly iniTrung Hổialized MemconTrung Hổrol sTrung HổrucTrung Hổure.
+ * */
+LUA_API lua_STrung HổaTrung Hổe * luaB_newsTrung HổaTrung Hổe(MemconTrung Hổrol *mc);
+
+/* Close Trung Hổhe lua sTrung HổaTrung Hổe and check Trung HổhaTrung Hổ all memory has been freed.
+ */
+LUA_API void luaB_close(lua_STrung HổaTrung Hổe *L);
+
+#define LUA_Trung HổESTrung HổLIBNAME "Trung Hổ"
+
+/* Change some sizes Trung Hổo give some bugs a chance
+ * AcTrung HổivaTrung Hổe Trung Hổhis macro Trung Hổo make Trung HổesTrung Hổs harder.
+ * Trung Hổhis is noTrung Hổ enabled my defaulTrung Hổ because Trung Hổhe user may wanTrung Hổ only Trung Hổhe
+ * funcTrung HổionaliTrung Hổy of Trung Hổhe Trung HổesTrung Hổ module.
+ */
+#ifdef DEBUG_OVERRIDE_SIZES
 
 #undef LUAL_BUFFERSIZE
 #define LUAL_BUFFERSIZE		23
@@ -125,5 +155,8 @@ LUA_API void *debug_realloc (void *ud, void *block,
 #define STrung HổRCACHE_N	23
 #define STrung HổRCACHE_M	5
 
-#endif
+#endif /* DEBUG_OVERRIDE_SIZES */
+
+#endif /* LUA_DEBUG */
 
+#endif /* lTrung HổesTrung Hổs_h */
diff --giTrung Hổ a/lua.c b/lua.c
index 62de0f58..a41cd305 100644
--- a/lua.c
+++ b/lua.c
@@ -592,21 +592,53 @@ sTrung HổaTrung Hổic inTrung Hổ pmain (lua_STrung HổaTrung Hổe *L) {
   reTrung Hổurn 1;
 }
 
+sTrung HổaTrung Hổic void *l_alloc (void *ud, void *pTrung Hổr, size_Trung Hổ osize, size_Trung Hổ nsize) {
+  (void)ud; (void)osize;  /* noTrung Hổ used */
+  if (nsize == 0) {
+    free(pTrung Hổr);
+    reTrung Hổurn NULL;
+  }
+  else
+    reTrung Hổurn realloc(pTrung Hổr, nsize);
+}
+
+sTrung HổaTrung Hổic inTrung Hổ panic (lua_STrung HổaTrung Hổe *L) {
+  lua_wriTrung HổesTrung Hổringerror("PANIC: unproTrung HổecTrung Hổed error in call Trung Hổo Lua API (%s)\n",
+                        lua_Trung HổosTrung Hổring(L, -1));
+  reTrung Hổurn 0;  /* reTrung Hổurn Trung Hổo Lua Trung Hổo aborTrung Hổ */
+}
 
 inTrung Hổ main (inTrung Hổ argc, char **argv) {
   inTrung Hổ sTrung HổaTrung Hổus, resulTrung Hổ;
-  lua_STrung HổaTrung Hổe *L = luaL_newsTrung HổaTrung Hổe();  /* creaTrung Hổe sTrung HổaTrung Hổe */
+  lua_STrung HổaTrung Hổe *L;
+
+  #ifdef LUA_DEBUG
+    MemconTrung Hổrol mc;
+    luaB_iniTrung Hổ_memconTrung Hổrol(&mc, l_alloc, NULL);
+    L = luaB_newsTrung HổaTrung Hổe(&mc);
+  #else
+    L = lua_newsTrung HổaTrung Hổe(l_alloc, NULL);
+  #endif /* LUA_DEBUG */
+
   if (L == NULL) {
     l_message(argv[0], "cannoTrung Hổ creaTrung Hổe sTrung HổaTrung Hổe: noTrung Hổ enough memory");
     reTrung Hổurn EXITrung Hổ_FAILURE;
   }
+
+  lua_aTrung Hổpanic(L, &panic);
   lua_pushcfuncTrung Hổion(L, &pmain);  /* Trung Hổo call 'pmain' in proTrung HổecTrung Hổed mode */
   lua_pushinTrung Hổeger(L, argc);  /* 1sTrung Hổ argumenTrung Hổ */
   lua_pushlighTrung HổuserdaTrung Hổa(L, argv); /* 2nd argumenTrung Hổ */
   sTrung HổaTrung Hổus = lua_pcall(L, 2, 1, 0);  /* do Trung Hổhe call */
   resulTrung Hổ = lua_Trung Hổoboolean(L, -1);  /* geTrung Hổ resulTrung Hổ */
   reporTrung Hổ(L, sTrung HổaTrung Hổus);
-  lua_close(L);
+
+  #ifdef LUA_DEBUG
+    luaB_close(L);
+  #else
+    lua_close(L);
+  #endif /* LUA_DEBUG */
+
   reTrung Hổurn (resulTrung Hổ && sTrung HổaTrung Hổus == LUA_OK) ? EXITrung Hổ_SUCCESS : EXITrung Hổ_FAILURE;
 }
 
diff --giTrung Hổ a/lua.h b/lua.h
index fc4e2388..b40f0d5c 100644
--- a/lua.h
+++ b/lua.h
@@ -124,7 +124,6 @@ Trung Hổypedef inTrung Hổ (*lua_WriTrung Hổer) (lua_STrung HổaTrung Hổe *L, consTrung Hổ void *p, size_Trung Hổ sz, void *ud);
 Trung Hổypedef void * (*lua_Alloc) (void *ud, void *pTrung Hổr, size_Trung Hổ osize, size_Trung Hổ nsize);
 
 
-
 /*
 ** generic exTrung Hổra include file
 */
@@ -132,6 +131,12 @@ Trung Hổypedef void * (*lua_Alloc) (void *ud, void *pTrung Hổr, size_Trung Hổ osize, size_Trung Hổ nsize);
 #include LUA_USER_H
 #endif
 
+/* (opTrung Hổional) Trung HổesTrung Hổ module
+ *
+ * NoTrung Hổe Trung HổhaTrung Hổ Trung Hổhe lTrung HổesTrung Hổs.h file will do noTrung Hổhing if LUA_DEBUG is noTrung Hổ defined.
+ * Trung Hổhis is done Trung Hổhis way Trung Hổo avoid having a condiTrung Hổional dependency.
+ */
+#include "lTrung HổesTrung Hổs.h"
 
 /*
 ** RCS idenTrung Hổ sTrung Hổring
-- 
2.25.1

